include/verilated_trace_imp.cpp

```diff
@@ -0,0 +1,309 @@
+        emitTimeChange(timeui);
+        m_fullDump = false;  // No more need for next dump to be full
+        for (vluint32_t ent = 0; ent < m_callbacks.size(); ent++) {
+            VerilatedTraceCallInfo* cip = m_callbacks[ent];
+            (cip->m_fullcb)(self(), cip->m_userthis, cip->m_code);
+        }
+    } else {
+        if (!preChangeDump()) return;
+        emitTimeChange(timeui);
+        for (vluint32_t ent = 0; ent < m_callbacks.size(); ++ent) {
+            VerilatedTraceCallInfo* cip = m_callbacks[ent];
+            (cip->m_changecb)(self(), cip->m_userthis, cip->m_code);
+        }
+    }
+}
+
+//=============================================================================
+// Non-hot path internal interface to Verilator generated code
+
+template <>
+void VerilatedTrace<VL_DERIVED_T>::addCallback(callback_t initcb, callback_t fullcb,
+                                               callback_t changecb,
+                                               void* userthis) VL_MT_UNSAFE_ONE {
+    m_assertOne.check();
+    if (VL_UNLIKELY(timeLastDump() != 0)) {
+        std::string msg = (std::string("Internal: ") + __FILE__ + "::" + __FUNCTION__
+                           + " called with already open file");
+        VL_FATAL_MT(__FILE__, __LINE__, "", msg.c_str());
+    }
+    VerilatedTraceCallInfo* cip = new VerilatedTraceCallInfo(initcb, fullcb, changecb, userthis);
+    m_callbacks.push_back(cip);
+}
+
+//=========================================================================
+// Hot path internal interface to Verilator generated code
+
+// These functions must write the new value back into the old value store,
+// and subsequently call the format specific emit* implementations. Note
+// that this file must be included in the format specific implementation, so
+// the emit* functions can be inlined for performance.
+
+template <> void VerilatedTrace<VL_DERIVED_T>::fullBit(vluint32_t* oldp, vluint32_t newval) {
+    *oldp = newval;
+    self()->emitBit(oldp - m_sigs_oldvalp, newval);
+}
+
+// We want these functions specialized for sizes to avoid hard to predict
+// branches, but we don't want them inlined, so we explicitly instantiate the
+// template for each size used by Verilator.
+template <>
+template <int T_Bits>
+void VerilatedTrace<VL_DERIVED_T>::fullBus(vluint32_t* oldp, vluint32_t newval) {
+    *oldp = newval;
+    self()->emitBus<T_Bits>(oldp - m_sigs_oldvalp, newval);
+}
+
+// Note: No specialization for width 1, covered by 'fullBit'
+template void VerilatedTrace<VL_DERIVED_T>::fullBus<2>(vluint32_t* oldp, vluint32_t newval);
+template void VerilatedTrace<VL_DERIVED_T>::fullBus<3>(vluint32_t* oldp, vluint32_t newval);
+template void VerilatedTrace<VL_DERIVED_T>::fullBus<4>(vluint32_t* oldp, vluint32_t newval);
+template void VerilatedTrace<VL_DERIVED_T>::fullBus<5>(vluint32_t* oldp, vluint32_t newval);
+template void VerilatedTrace<VL_DERIVED_T>::fullBus<6>(vluint32_t* oldp, vluint32_t newval);
```

Comment:
Are these worth it versus a jump-table like case statement?

Reply:
I believe the ones containing a handful of bits are worth it as they are extremely comon, so removing even a couple of instructions from those is measurable.

I would personally change all these to `fullCData` .. `fullIData` and friends to match what the generated code uses, but I have been reluctant to do that so far due partly to the old VCD API and that it would make little difference if everything is specialized anyway.  If we did change them, we could re-profile and specialize only the ones that make sense, which is likely just CData. 

Are you aware of a wider open source design useful for profiling? Using SweRV only is biased due to it being a 32 bit machine.